Add MSSQL support

phanikumv opened the pull request:
# Description
## What is the current behavior?
<!-- Please describe the current behavior that you are modifying. -->
astro-sdk operators currently do not work with mssql


closes: #1109 


## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

- Add mssql support


## Does this introduce a breaking change?
No

### Checklist
- [x] Created tests which fail without the change (if possible)
- [x] Extended the README / documentation, if necessary

utkarsharma2 commented on python-sdk/src/astro/databases/base.py:
```diff
@@ -345,7 +345,10 @@ def drop_table(self, table: BaseTable) -> None:
         """
         statement = self._drop_table_statement.format(self.get_table_qualified_name(table))
-        self.run_sql(statement)
+        if self.sql_type == "mssql":
```
@phanikumv Why do we need this?

utkarsharma2 replied:
Also, if this is needed we should move it to databases/mssql.py by overriding `drop_table` method.

phanikumv replied:
we need this to be able to add autocommit flag for certain sql commands. Will move it to `databases/mssql.py`

utkarsharma2 commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,359 @@
+    END
+    """
+
+    _queries_requiring_autocommit: list = ["alter", "select * into"]
```
@phanikumv Why do we need this only for these queries?

phanikumv replied:
because alter , select * into, drop table are considered as DDL(they change the system catalog). The pymssql's sqlalchemy considers autocommit as False by default which causes these DDL's to not execute

phanikumv replied:
Added comment explaining this in [88008cd](https://github.com/astronomer/astro-sdk/pull/1538/commits/88008cd4593127db9111768fa2dbef51502d955d)

utkarsharma2 commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,359 @@
+
+    _queries_requiring_autocommit: list = ["alter", "select * into"]
+
+    def __init__(self, conn_id: str = DEFAULT_CONN_ID, table: BaseTable | None = None):
```
@phanikumv Are there any load_options required for this db?

phanikumv replied:
added `load_options`

utkarsharma2 commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,359 @@
+        schema = None
+        if table and table.metadata:
+            schema = table.metadata.schema
+        # elif self.hook.schema:
+        #     schema = self.hook.schema
```
```suggestion
```

phanikumv replied:
Done in [88008cd](https://github.com/astronomer/astro-sdk/pull/1538/commits/88008cd4593127db9111768fa2dbef51502d955d)

utkarsharma2 commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,359 @@
+        :param table: The table we want to retrieve the qualified name for.
+        """
+        qualified_name_lists = [
+            table.metadata.database,
```
is Database name required?

phanikumv replied:
yes it is required, the fully qualified name for a table will be `<db>.<schema>.<tbl>`

utkarsharma2 commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,359 @@
+            schema = table.metadata.schema
+        # elif self.hook.schema:
+        #     schema = self.hook.schema
+        return bool(inspector.dialect.has_table(self.connection, table.name, schema=schema))
```
a fully qualified name is not working here? 
like: return bool(inspector.dialect.has_table(self.connection, table_qualified_name))

phanikumv replied:
It doesnt work because the database would contain the name as `abc` instead of `db.schema.abc`.

sunank200 commented on python-sdk/tests_integration/sql/operators/test_load_file.py:
```diff
@@ -761,6 +820,48 @@ def test_load_file(sample_dag, database_table_fixture, file_type):
+)
+@pytest.mark.parametrize("file_type", ["parquet", "ndjson", "json", "csv"])
+def test_load_file_for_mssql(sample_dag, database_table_fixture, file_type):
+    db, test_table = database_table_fixture
```
Doc-string would be good

phanikumv replied:
added docstring

sunank200 commented on python-sdk/tests_integration/sql/operators/test_load_file.py:
```diff
@@ -821,6 +925,25 @@ def validate(table: Table):
 
+@pytest.mark.integration
+def test_populate_table_metadata_mssql(sample_dag):
+    """
+    Test default populating of table fields in load_fil op.
+    """
```
```suggestion
    """Test default populating of table fields in load_fil op."""
```

phanikumv replied:
Fixed

sunank200 reviewed (commented):
Added comments but i think we should also add few integration test as part of example DAG for mysql as well. WDYT?

phanikumv commented on python-sdk/src/astro/databases/base.py:
```diff
@@ -345,7 +345,10 @@ def drop_table(self, table: BaseTable) -> None:
         statement = self._drop_table_statement.format(self.get_table_qualified_name(table))
-        self.run_sql(statement)
+        if self.sql_type == "mssql":
+            self.run_sql(statement, autocommit=True)
```
Done in [88008cd](https://github.com/astronomer/astro-sdk/pull/1538/commits/88008cd4593127db9111768fa2dbef51502d955d)

phanikumv commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,359 @@
+        super().__init__(conn_id)
+        self.table = table
+
+    @property
```
changed to use `@cached_property` - just to note tha the other database implementations too have this as `@property`

utkarsharma2 commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,378 @@
+            sql = kwargs.get("sql_statement")  # type: ignore
+
+        # We need to set autocommit=True for specific queries
+        if isinstance(sql, str):
```
@phanikumv This looks complex, can move the logic to decide the auto-commit required into a separate function?
`is_auto_commit_needed(query) -> bool`

utkarsharma2 replied:
Suggestion: return self.connection.execute(sqlalchemy.text(sql).execution_options(autocommit=is_autocommit_needed(sql)), parameters)

phanikumv replied:
this is changed now

utkarsharma2 commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,378 @@
+        self,
+        sql: str | ClauseElement = "",
+        parameters: dict | None = None,
+        autocommit: bool = False,
```
@phanikumv  This signature doesn't match with the base database class, it would be good if we can make `autocommit` part of the base class as well.

phanikumv replied:
done

utkarsharma2 commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,378 @@
+                return result
+        else:
+            # this is used for append
+            result = self.connection.execute(sql, parameters)
```
@phanikumv We don't need auto-commit options when the query is not string?

phanikumv replied:
no we don't as its a compiled sql statement that comes into the else block

utkarsharma2 commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,359 @@
+        """
+        if schema and not self.schema_exists(schema):
+            statement = self._create_schema_statement.format(schema)
+            self.run_sql(statement, autocommit=True)
```
or extend existing test case.

phanikumv replied:
added test for this

utkarsharma2 commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,378 @@
+            result = self.connection.execute(sql, parameters)
+            return result
+
+    def create_schema_if_needed(self, schema: str | None) -> None:
```
@phanikumv Are we overriding `create_schema_if_needed` for `self.run_sql(statement, autocommit=True`)? 

If so, maybe we can abstract away `autocommit` required logic in a different function?

phanikumv replied:
why do we need further abstraction? you think overriding `create_schema_if_needed` for `self.run_sql(statement, autocommit=True)` is not a good approach?

phanikumv replied:
overriding `create_schema_if_needed` is needed for two reasons:

- the `create_schema_statement` is different between base class and the mssql based implementation.
- self.run_sql(statement, autocommit=True)

kaxil commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,382 @@
+
+    def table_exists(self, table: BaseTable) -> bool:
+        """
+        Check if a table exists in the database
+        :param table: Details of the table we want to check that exists
```
```suggestion
        Check if a table exists in the database

        :param table: Details of the table we want to check that exists
```

phanikumv replied:
fixed

kaxil commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,382 @@
+    @staticmethod
+    def get_table_qualified_name(table: BaseTable) -> str:
+        """
+        Return the table qualified name.
+        :param table: The table we want to retrieve the qualified name for.
```
```suggestion
        Return the table qualified name.

        :param table: The table we want to retrieve the qualified name for.
```

phanikumv replied:
fixed

kaxil commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,382 @@
+
+    def schema_exists(self, schema) -> bool:
+        """
+        Checks if a schema exists in the database
+        :param schema: DB Schema - a namespace that contains named objects like (tables, functions, etc)
```
```suggestion
        Checks if a schema exists in the database

        :param schema: DB Schema - a namespace that contains named objects like (tables, functions, etc)
```

phanikumv replied:
Fixed

kaxil commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,382 @@
+        """
+        Return the results to running a SQL statement.
+        Whenever possible, this method should be implemented using Airflow Hooks,
+        since this will simplify the integration with Async operators.
+        :param sql: Contains SQL query to be run against database
```
```suggestion
        since this will simplify the integration with Async operators.

        :param sql: Contains SQL query to be run against database
```

phanikumv replied:
fixed

kaxil commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,382 @@
+    def create_schema_if_needed(self, schema: str | None) -> None:
+        """
+        This function checks if the expected schema exists in the database. If the schema does not exist,
+        it will attempt to create it.
+        :param schema: DB Schema - a namespace that contains named objects like (tables, functions, etc)
```
```suggestion
        it will attempt to create it.

        :param schema: DB Schema - a namespace that contains named objects like (tables, functions, etc)
```

phanikumv replied:
fixed

kaxil commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,382 @@
+        parameters: dict | None = None,
+    ) -> None:
+        """
+        Export the result rows of a query statement into another table.
+        :param statement: SQL query statement
```
```suggestion
        Export the result rows of a query statement into another table.

        :param statement: SQL query statement
```

phanikumv replied:
fixed

kaxil commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,382 @@
+    ) -> None:  # skipcq PYL-W0613
+        """
+        Create a table with the dataframe's contents.
+        If the table already exists, append or replace the content, depending on the value of `if_exists`.
+        :param source_dataframe: Local or remote filepath
```
```suggestion
        If the table already exists, append or replace the content, depending on the value of `if_exists`.

        :param source_dataframe: Local or remote filepath
```

kaxil replied:
and same everywhere else. Check whitespace error code in https://www.pydocstyle.org/en/stable/error_codes.html or https://google.github.io/styleguide/pyguide.html#381-docstrings

phanikumv replied:
fixed

tatiana commented on python-sdk/tests/data/sample_without_unicode.csv:
```diff
@@ -0,0 +1,3 @@
+id,name
+1,First
+2,Second
```
Are there any issues with using our existing examples with Unicode in MSSQL?

phanikumv replied:
yes we get garbled text into the database when we try to insert the unicode data. So the tests with unicode data will fail

phanikumv replied:
will add a note in the documentation that unicode isnt supported in this implementation

phanikumv commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,385 @@
+        initialization, but it uses it as a database in the connection to Mssql:
+        """
+        # TODO: Change airflow MssqlHook to fetch database and schema separately
+        database = self.hook.get_connection(self.conn_id).schema
```
it sets the schema sent by user as database

```
    def get_conn(self) -> pymssql.connect:
        """Returns a mssql connection object"""
        conn = self.get_connection(self.mssql_conn_id)  # type: ignore[attr-defined]
        conn = pymssql.connect(
            server=conn.host,
            user=conn.login,
            password=conn.password,
            database=self.schema or conn.schema,  <-------------
            port=conn.port,
        )

pankajastro commented on logs/scheduler/latest:
```diff
@@ -0,0 +1 @@
+/Users/phanikv/Documents/wip/astro-sdk/astro-sdk/logs/scheduler/2023-01-16
```
do we need to check-in this?

pankajastro replied:
looks your local machine file path

phanikumv replied:
my bad, I've now removed this file

pankajastro commented on webserver_config.py:
```diff
@@ -0,0 +1,130 @@
+#
```
just curious why do we need this file?

pankajastro commented on python-sdk/tests_integration/sql/operators/transform/test_transform.py:
```diff
@@ -309,6 +492,20 @@ def top_five_animations(input_table: Table) -> str:
+    assert task.operator.outlets == [output_table]
+
+
 @pytest.mark.skipif(DATASET_SUPPORT, reason="Inlets/Outlets will only be added for Airflow >= 2.4")
```
should we add `@pytest.mark.integration`?

phanikumv replied:
fixed

pankajastro commented on python-sdk/tests_integration/sql/operators/transform/test_transform.py:
```diff
@@ -249,6 +406,32 @@ def select(input_table: Table):
     test_utils.run_dag(sample_dag)
 
 
+def test_transform_using_table_metadata_mssql(sample_dag):
```
do we need to add `@pytest.mark.integration`?

phanikumv replied:
fixed

phanikumv commented on python-sdk/pyproject.toml:
```diff
@@ -90,6 +90,10 @@ openlineage = ["openlineage-airflow>=0.17.0"]
 databricks = ["databricks-cli",
     "apache-airflow-providers-databricks"]
 
+mssql = [
```
@tatiana you meant to add `apache-airflow-providers-microsoft-mssql` into the azure block ?

sunank200 replied:
I think `apache-airflow-providers-microsoft-mssq` would make sense to be in mssql block itself as it represents the database

phanikumv replied:
yes even I think it just needs to be in `mssql` block. But would wait for @tatiana's feedback

phanikumv commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,359 @@
+        if row_limit > -1:
+            statement = f"SELECT TOP {row_limit} * FROM {self.get_table_qualified_name(table)}"
+        response = self.run_sql(statement)
+        return response.fetchall()  # type: ignore
```
@sunank200  It will result in below error otherwise. This  `# type: ignore` is present in `BaseDatabase` class also. If we want this to be fixed, then we need to do this across the repo.

```src/astro/databases/mssql.py: note: In member "fetch_all_rows" of class "MssqlDatabase":
src/astro/databases/mssql.py:237: error: Returning Any from function declared
to return "List[Any]"  [no-any-return]
            return response.fetchall()
```

sunank200 reviewed (commented):
Overall the PR seems in much better shape now. Added few comments. Also following things would be required:
- documentation changes for mssql
- Example DAG in loadfile and export file

kaxil commented on python-sdk/pyproject.toml:
```diff
@@ -21,6 +21,7 @@ dependencies = [
     "attrs>=20.0",
     "pandas>=1.3.4,<2.0.0", # Pinning it to <2.0.0 to avoid breaking changes
     "pyarrow",
+    "pyodbc",
```
Why do we need to make `pyodbc` as a required core dep here?

Do we need it for core or is this just needed for mssql?

phanikumv replied:
I added it just to test, removed it later

kaxil commented on python-sdk/pyproject.toml:
```diff
@@ -90,6 +91,10 @@ openlineage = ["openlineage-airflow>=0.17.0"]
     "apache-airflow-providers-databricks"]
 
+mssql = [
+    "apache-airflow-providers-microsoft-mssql",
```
```suggestion
    "apache-airflow-providers-microsoft-mssql>=3.2",
```

kaxil commented on python-sdk/pyproject.toml:
```diff
@@ -107,7 +111,8 @@ all = [
     "apache-airflow-providers-microsoft-azure",
-    "azure-storage-blob"
+    "azure-storage-blob",
+    "apache-airflow-providers-microsoft-mssql"
```
```suggestion
    "apache-airflow-providers-microsoft-mssql>=3.2"
```

kaxil replied:
We should have the same deps in both places

phanikumv replied:
yes, made the changes

feluelle commented on python-sdk/tests_integration/databases/test_all_databases.py:
```diff
@@ -148,6 +182,29 @@ def test_fetch_all_rows(mock_run_sql, database_table_fixture, row_count):
+    db.fetch_all_rows(table, row_count)
+    select_statements = [m.args[0] for m in mock_run_sql.mock_calls if m.args and "SELECT" in m.args[0]]
+    assert len(select_statements) == 1
+    if row_count > -1:
+        assert f"TOP {row_count}" in select_statements[0]
+    else:
+        assert f"TOP {row_count}" not in select_statements[0]
```
According to `@pytest.mark.parametrize("row_count", [0, 100])` the row count will always be above -1. Why do we need the if/else?

phanikumv replied:
fixed it

phanikumv commented on python-sdk/noxfile.py:
```diff
@@ -36,15 +36,11 @@ def test(session: nox.Session, airflow) -> None:
         # install apache-airflow-providers-sftp==4.0.0 since it is the compatible version
         # to run sftp example dag to load file with airflow 2.2.5
-        session.install("apache-airflow-providers-sftp==4.0.0")
+        # session.install("apache-airflow-providers-sftp==4.0.0")
```
fixed now

phanikumv commented on python-sdk/src/astro/databases/mssql.py:
```diff
@@ -0,0 +1,359 @@
+        Checks if a schema exists in the database
+        :param schema: DB Schema - a namespace that contains named objects like (tables, functions, etc)
+        """
+        schema_result = self.hook.run(
```
added exception handling

phanikumv commented on python-sdk/noxfile.py:
```diff
@@ -36,15 +36,11 @@ def test(session: nox.Session, airflow) -> None:
         # install apache-airflow-providers-sftp==4.0.0 since it is the compatible version
         # to run sftp example dag to load file with airflow 2.2.5
-        session.install("apache-airflow-providers-sftp==4.0.0")
+        # session.install("apache-airflow-providers-sftp==4.0.0")
```
this is outdated, I have removed commented code

phanikumv commented on python-sdk/noxfile.py:
```diff
@@ -36,15 +36,11 @@ def test(session: nox.Session, airflow) -> None:
-        )
-        session.install(f"apache-airflow=={airflow}", "-c", constraints_url)
-        session.install("-e", ".[all,tests]", "-c", constraints_url)
+        session.install(f"apache-airflow=={airflow}", "-c", "tests/modified_constraint_file.txt")
```
have put the required comment to explain this

phanikumv commented on python-sdk/example_dags/example_load_file.py:
```diff
@@ -343,4 +344,13 @@
     )
     # [END load_file_example_25]
 
+    # [START load_file_example_26]
```
we are using this in `load_file.rst`

phanikumv commented on python-sdk/example_dags/example_mssql_transform.py:
```diff
@@ -0,0 +1,101 @@
+# [END transform_example_3]
+
+
+# [START transform_example_4]
```
removed

phanikumv commented on python-sdk/example_dags/example_mssql_transform.py:
```diff
@@ -0,0 +1,101 @@
+# [END transform_example_2]
+
+
+# [START transform_example_3]
```
removed

phanikumv commented on python-sdk/example_dags/example_mssql_transform.py:
```diff
@@ -0,0 +1,101 @@
+# [END transform_example_1]
+
+
+# [START transform_example_2]
```
removed

phanikumv commented on python-sdk/example_dags/example_mssql_transform.py:
```diff
@@ -0,0 +1,101 @@
+LAST_ONE_DF = pd.DataFrame(data={"title": ["Random movie"], "rating": [121]})
+
+
+# [START transform_example_1]
```
removed

phanikumv commented:
> Added comments but i think we should also add few integration test as part of example DAG for mysql as well. WDYT?

This is completed

phanikumv commented:
> Overall the PR seems in much better shape now. Added few comments. Also following things would be required:
> 
> * documentation changes for mssql
> * Example DAG in loadfile and export file



> Overall the PR seems in much better shape now. Added few comments. Also following things would be required:
> 
> * documentation changes for mssql
> * Example DAG in loadfile and export file

Added documentation and example DAG
